spec/formatting.md

```diff
@@ -18,3 +18,192 @@ the local variable takes precedence.
 
 It is an error for a local variable definition to
 refer to a local variable that's defined after it in the message.
+
+## Error Handling
+
+During the formatting of a message,
+various errors may be encountered.
+These are divided into the following categories:
+
+- **Syntax errors** occur when the syntax representation of a message is not well-formed.
+
+  Example invalid messages resulting in a Syntax error:
+
+  ```
+  {Missing end brace
+  ```
+
+  ```
+  {Unknown {?placeholder}}
+  ```
+
+  ```
+  let $var = {(no message body)}
+  ```
+
+- **Data Model errors** occur when a message is invalid due to
+  violating one of the following semantic requirements on its structure:
+
+  - **Variant Key Mismatch errors** occur when the number of keys on a Variant
+    does not equal the number of Selectors.
+
+    Example invalid messages resulting in a Variant Key Mismatch error:
+
+    ```
+    match {$one}
+    when 1 2 {Too many}
+    when * {Otherwise}
+    ```
+
+    ```
+    match {$one} {$two}
+    when 1 2 {Two keys}
+    when * {Too few}
+    when * * {Otherwise}
+    ```
+
+  - **Missing Fallback Variant errors** occur when the message
+    does not include a Variant with only catch-all keys.
+
+    Example invalid messages resulting in a Missing Fallback Variant error:
+
+    ```
+    match {$one}
+    when 1 {Value is one}
+    when 2 {Value is two}
+    ```
+
+    ```
+    match {$one} {$two}
+    when 1 * {First is one}
+    when * 1 {Second is one}
+    ```
+
+- **Resolution errors** occur when the runtime value of a part of a message
+  cannot be determined.
+
+  - **Unresolved Variable errors** occur when a variable reference cannot be resolved.
+
+    For example, attempting to format either of the following messages
+    must result in an Unresolved Variable error if done within a context that
+    does not provide for the variable reference `$var` to be successfully resolved:
+
+    ```
+    {The value is {$var}.}
+    ```
+
+    ```
+    match {$var}
```

Comment:
Something that I don't see addressed explicitly: should `match {$var}` here emit both an Unresolved Variable error and a Selector error? More generally: can errors chain?

Reply:
Regarding the specific case, this is in fact [explicitly addressed](https://github.com/unicode-org/message-format-wg/pull/320/files#diff-a77dffbd4837df97c09741f8c339a41e3730ad66fe20358ff9fa27b9bf38d839R161-R163):
> When an error occurs within a Selector, the selector must not match any VariantKey other than the catch-all `*` and a Selector error is emitted.

However, I now think that text ought to include some explicit clarifier that an implementation which only supports one error can just emit the earlier error instead.

Regarding the general case: yes, errors may chain. What that means in practice ought to be implementation-specific.